std/machines/hash/keccakf16.asm

```diff
@@ -0,0 +1,608 @@
+    //             local.c[x][z].into(),
+    //             local.c[(x + 4) % 5][z].into(),
+    //             local.c[(x + 1) % 5][(z + 63) % 64].into(),
+    //         );
+    //         let c_prime = local.c_prime[x][z];
+    //         builder.assert_eq(c_prime, xor);
+    //     }
+    // }
+
+    array::map(c, |i| force_bool(i));
+    
+    let andn: expr, expr -> expr = |a, b| (1 - a) * b;
+    let xor: expr, expr -> expr = |a, b| a + b - 2*a*b;
+    let xor3: expr, expr, expr -> expr = |a, b, c| xor(xor(a, b), c);
+    // a b c xor3
+    // 0 0 0  0
+    // 0 0 1  1
+    // 0 1 0  1
+    // 0 1 1  0
+    // 1 0 0  1
+    // 1 0 1  0
+    // 1 1 0  0
+    // 1 1 1  1
+    array::new(320, |i| {
+        let x = i / 64;
+        let z = i % 64;
+        c_prime[i] = xor3(
+            c[i], 
+            c[((x + 4) % 5) * 64 + z], 
+            c[((x + 1) % 5) * 64 + ((z + 63) % 64)]
+        )
+    });
+
+    // // Check that the input limbs are consistent with A' and D.
+    // // A[x, y, z] = xor(A'[x, y, z], D[x, y, z])
+    // //            = xor(A'[x, y, z], C[x - 1, z], C[x + 1, z - 1])
+    // //            = xor(A'[x, y, z], C[x, z], C'[x, z]).
+    // // The last step is valid based on the identity we checked above.
+    // // It isn't required, but makes this check a bit cleaner.
+    // for y in 0..5 {
+    //     for x in 0..5 {
+    //         let get_bit = |z| {
+    //             let a_prime: AB::Var = local.a_prime[y][x][z];
+    //             let c: AB::Var = local.c[x][z];
+    //             let c_prime: AB::Var = local.c_prime[x][z];
+    //             xor3_gen::<AB::Expr>(a_prime.into(), c.into(), c_prime.into())
+    //         };
+
+    //         for limb in 0..U64_LIMBS {
+    //             let a_limb = local.a[y][x][limb];
+    //             let computed_limb = (limb * BITS_PER_LIMB..(limb + 1) * BITS_PER_LIMB) // bigger address correspond to more significant bit
+    //                 .rev()
+    //                 .fold(AB::Expr::zero(), |acc, z| {
+    //                     builder.assert_bool(local.a_prime[y][x][z]);
+    //                     acc.double() + get_bit(z)
+    //                 });
+    //             builder.assert_eq(computed_limb, a_limb);
+    //         }
+    //     }
+    // }
+
+    let bits_to_value_be: expr[] -> expr = |bits_be| array::fold(bits_be, 0, |acc, e| (acc * 2 + e));
+
+    array::map(a_prime, |i| force_bool(i));
+    array::new(100, |i| {
+        let y = i / 20;
+        let x = (i / 4) % 5;
+        let limb = i % 4;
+        let get_bit: int -> expr = |z| xor3(a_prime[y * 320 + x * 64 + z], c[x * 64 + z], c_prime[x * 64 + z]);
+        let limb_bits_be: expr[] = array::reverse(array::new(16, |z| get_bit(limb * 16 + z)));
+        a[i] = bits_to_value_be(limb_bits_be)
+    });
+
+    // // xor_{i=0}^4 A'[x, i, z] = C'[x, z], so for each x, z,
+    // // diff * (diff - 2) * (diff - 4) = 0, where
+    // // diff = sum_{i=0}^4 A'[x, i, z] - C'[x, z]
```

Comment:
This is basically saying that if five bits xor to one bit, then the difference between the sum of the five bits and the one bit is either 0, 2, or 4.

For example, 0 = 1 xor 1 xor 0 xor 0 xor 0, so diff = (1 + 1 + 0 + 0 + 0) - 0 = 2.

It can be proven mathematically that these two statements are equivalent.

Reply:
Yes, I understood!

I think it's pretty smart, because now 5 XORs are computed with one degree-3 constraint, whereas constraining it "naively" (repeatedly applying `xor = |a, b| a + b - 2*a*b`) would yield a degree-6 constraint, which would be above the degree bound and we'd have to add witness columns to store intermediate results.